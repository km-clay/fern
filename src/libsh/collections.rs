use std::collections::VecDeque;

pub trait VecDequeAliases<T> {
	fn fpop(&mut self) -> Option<T>;
	fn fpush(&mut self, value: T);
	fn bpop(&mut self) -> Option<T>;
	fn bpush(&mut self, value: T);
	fn to_vec(self) -> Vec<T>;
}

impl<T> VecDequeAliases<T> for VecDeque<T> {
	/// Alias for pop_front()
	fn fpop(&mut self) -> Option<T> {
		self.pop_front()
	}
	/// Alias for push_front()
	fn fpush(&mut self, value: T) {
		self.push_front(value);
	}
	/// Alias for pop_back()
	fn bpop(&mut self) -> Option<T> {
		self.pop_back()
	}
	/// Alias for push_back()
	fn bpush(&mut self, value: T) {
		self.push_back(value);
	}
	/// Just turns the deque into a vector
	fn to_vec(mut self) -> Vec<T> {
		let mut vec = vec![];
		while let Some(item) = self.fpop() {
			vec.push(item)
		}
		vec
	}
}
