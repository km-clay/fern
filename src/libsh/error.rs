use std::{fmt::Display, str::FromStr};

use crate::{parse::lex::Span, prelude::*};

pub type ShResult<'s,T> = Result<T,ShErr<'s>>;

#[derive(Debug)]
pub enum ShErr<'s> {
	Simple { kind: ShErrKind, msg: String },
	Full { kind: ShErrKind, msg: String, span: Span<'s> }
}

impl<'s> ShErr<'s> {
	pub fn simple(kind: ShErrKind, msg: impl Into<String>) -> Self {
		let msg = msg.into();
		Self::Simple { kind, msg }
	}
	pub fn full(kind: ShErrKind, msg: impl Into<String>, span: Span<'s>) -> Self {
		let msg = msg.into();
		Self::Full { kind, msg, span }
	}
	pub fn unpack(self) -> (ShErrKind,String,Option<Span<'s>>) {
		match self {
			ShErr::Simple { kind, msg } => (kind,msg,None),
			ShErr::Full { kind, msg, span } => (kind,msg,Some(span))
		}
	}
	pub fn with_span(sherr: ShErr, span: Span<'s>) -> Self {
		let (kind,msg,_) = sherr.unpack();
		Self::Full { kind, msg, span }
	}
}

impl<'s> Display for ShErr<'s> {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Self::Simple { msg, kind: _ } => writeln!(f, "{}", msg),
			Self::Full { msg, kind: _, span: _ } => writeln!(f, "{}", msg)
		}
	}
}

impl<'s> From<std::io::Error> for ShErr<'s> {
	fn from(_: std::io::Error) -> Self {
		let msg = std::io::Error::last_os_error();
		ShErr::simple(ShErrKind::IoErr, msg.to_string())
	}
}

impl<'s> From<std::env::VarError> for ShErr<'s> {
	fn from(value: std::env::VarError) -> Self {
		ShErr::simple(ShErrKind::InternalErr, &value.to_string())
	}
}

impl<'s> From<rustyline::error::ReadlineError> for ShErr<'s> {
	fn from(value: rustyline::error::ReadlineError) -> Self {
		ShErr::simple(ShErrKind::ParseErr, &value.to_string())
	}
}

impl<'s> From<Errno> for ShErr<'s> {
	fn from(value: Errno) -> Self {
		ShErr::simple(ShErrKind::Errno, &value.to_string())
	}
}

#[derive(Debug)]
pub enum ShErrKind {
	IoErr,
	SyntaxErr,
	ParseErr,
	InternalErr,
	ExecFail,
	ResourceLimitExceeded,
	BadPermission,
	Errno,
	FileNotFound,
	CmdNotFound,
	CleanExit,
	FuncReturn,
	LoopContinue,
	LoopBreak,
	Null
}
