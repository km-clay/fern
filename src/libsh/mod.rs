pub mod error;
pub mod term;
