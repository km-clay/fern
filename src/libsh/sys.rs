use std::fmt::Display;

use crate::prelude::*;

pub const SIG_EXIT_OFFSET: i32 = 128;

pub fn get_bin_path(command: &str, shenv: &ShEnv) -> Option<PathBuf> {
	let env = shenv.vars().env();
	let path_var = env.get("PATH")?;
	let mut paths = path_var.split(':');
	while let Some(raw_path) = paths.next() {
		let mut path = PathBuf::from(raw_path);
		path.push(command);
		//TODO: handle this unwrap
		if path.exists() {
			return Some(path)
		}
	}
	None
}

pub fn write_out(text: impl Display) -> ShResult<()> {
	write(borrow_fd(1), text.to_string().as_bytes())?;
	Ok(())
}

pub fn write_err(text: impl Display) -> ShResult<()> {
	write(borrow_fd(2), text.to_string().as_bytes())?;
	Ok(())
}

/// Return is `readpipe`, `writepipe`
/// Contains all of the necessary boilerplate for grabbing two pipe fds using libc::pipe()
pub fn c_pipe() -> Result<(RawFd,RawFd),Errno> {
	let mut pipes: [i32;2] = [0;2];
	let ret = unsafe { libc::pipe(pipes.as_mut_ptr()) };
	if ret < 0 {
		return Err(Errno::from_raw(ret))
	}
	Ok((pipes[0],pipes[1]))
}

pub fn execvpe(cmd: String, argv: Vec<String>, envp: Vec<String>) -> Result<(),Errno> {
	let cmd_raw = CString::new(cmd).unwrap();

	let argv = argv.into_iter().map(|arg| CString::new(arg).unwrap()).collect::<Vec<CString>>();
	let envp = envp.into_iter().map(|var| CString::new(var).unwrap()).collect::<Vec<CString>>();

	nix::unistd::execvpe(&cmd_raw, &argv, &envp).unwrap();
	Ok(())
}
