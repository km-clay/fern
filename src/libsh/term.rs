use std::{fmt::Display, ops::BitOr};

pub trait Styled: Sized + Display {
	fn styled<S: Into<StyleSet>>(self, style: S) -> String {
		let styles: StyleSet = style.into();
		let reset = Style::Reset;
		format!("{styles}{self}{reset}")
	}
}

impl<T: Display> Styled for T {}

/// Enum representing a single ANSI style
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Style {
	// Undoes all styles
	Reset,
	// Foreground Colors
	Black,
	Red,
	Green,
	Yellow,
	Blue,
	Magenta,
	Cyan,
	White,
	BrightBlack,
	BrightRed,
	BrightGreen,
	BrightYellow,
	BrightBlue,
	BrightMagenta,
	BrightCyan,
	BrightWhite,
	RGB(u8, u8, u8),  // Custom foreground color

	// Background Colors
	BgBlack,
	BgRed,
	BgGreen,
	BgYellow,
	BgBlue,
	BgMagenta,
	BgCyan,
	BgWhite,
	BgBrightBlack,
	BgBrightRed,
	BgBrightGreen,
	BgBrightYellow,
	BgBrightBlue,
	BgBrightMagenta,
	BgBrightCyan,
	BgBrightWhite,
	BgRGB(u8, u8, u8),  // Custom background color

	// Text Attributes
	Bold,
	Dim,
	Italic,
	Underline,
	Strikethrough,
	Reversed,
}

impl Display for Style {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Style::Reset => write!(f, "\x1b[0m"),

			// Foreground colors
			Style::Black => write!(f, "\x1b[30m"),
			Style::Red => write!(f, "\x1b[31m"),
			Style::Green => write!(f, "\x1b[32m"),
			Style::Yellow => write!(f, "\x1b[33m"),
			Style::Blue => write!(f, "\x1b[34m"),
			Style::Magenta => write!(f, "\x1b[35m"),
			Style::Cyan => write!(f, "\x1b[36m"),
			Style::White => write!(f, "\x1b[37m"),
			Style::BrightBlack => write!(f, "\x1b[90m"),
			Style::BrightRed => write!(f, "\x1b[91m"),
			Style::BrightGreen => write!(f, "\x1b[92m"),
			Style::BrightYellow => write!(f, "\x1b[93m"),
			Style::BrightBlue => write!(f, "\x1b[94m"),
			Style::BrightMagenta => write!(f, "\x1b[95m"),
			Style::BrightCyan => write!(f, "\x1b[96m"),
			Style::BrightWhite => write!(f, "\x1b[97m"),
			Style::RGB(r, g, b) => write!(f, "\x1b[38;2;{r};{g};{b}m"),

			// Background colors
			Style::BgBlack => write!(f, "\x1b[40m"),
			Style::BgRed => write!(f, "\x1b[41m"),
			Style::BgGreen => write!(f, "\x1b[42m"),
			Style::BgYellow => write!(f, "\x1b[43m"),
			Style::BgBlue => write!(f, "\x1b[44m"),
			Style::BgMagenta => write!(f, "\x1b[45m"),
			Style::BgCyan => write!(f, "\x1b[46m"),
			Style::BgWhite => write!(f, "\x1b[47m"),
			Style::BgBrightBlack => write!(f, "\x1b[100m"),
			Style::BgBrightRed => write!(f, "\x1b[101m"),
			Style::BgBrightGreen => write!(f, "\x1b[102m"),
			Style::BgBrightYellow => write!(f, "\x1b[103m"),
			Style::BgBrightBlue => write!(f, "\x1b[104m"),
			Style::BgBrightMagenta => write!(f, "\x1b[105m"),
			Style::BgBrightCyan => write!(f, "\x1b[106m"),
			Style::BgBrightWhite => write!(f, "\x1b[107m"),
			Style::BgRGB(r, g, b) => write!(f, "\x1b[48;2;{r};{g};{b}m"),

			// Text attributes
			Style::Bold => write!(f, "\x1b[1m"),
			Style::Dim => write!(f, "\x1b[2m"),  // New
			Style::Italic => write!(f, "\x1b[3m"),
			Style::Underline => write!(f, "\x1b[4m"),
			Style::Strikethrough => write!(f, "\x1b[9m"),  // New
			Style::Reversed => write!(f, "\x1b[7m"),
		}
	}
}

/// Struct representing a **set** of styles
#[derive(Debug, Default, Clone)]
pub struct StyleSet {
	styles: Vec<Style>,
}

impl StyleSet {
	pub fn new() -> Self {
		Self { styles: vec![] }
	}

	pub fn add(mut self, style: Style) -> Self {
		if !self.styles.contains(&style) {
			self.styles.push(style);
		}
		self
	}
}

impl Display for StyleSet {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		for style in &self.styles {
			style.fmt(f)?
		}
		Ok(())
	}
}

/// Allow OR (`|`) operator to combine multiple `Style` values into a `StyleSet`
impl BitOr for Style {
	type Output = StyleSet;

	fn bitor(self, rhs: Self) -> Self::Output {
		StyleSet::new().add(self).add(rhs)
	}
}

/// Allow OR (`|`) operator to combine `StyleSet` with `Style`
impl BitOr<Style> for StyleSet {
	type Output = StyleSet;

	fn bitor(self, rhs: Style) -> Self::Output {
		self.add(rhs)
	}
}

impl From<Style> for StyleSet {
	fn from(style: Style) -> Self {
		StyleSet::new().add(style)
	}
}
