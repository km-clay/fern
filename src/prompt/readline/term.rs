use std::{env, fmt::{Debug, Write}, io::{BufRead, BufReader, Read}, iter::Peekable, os::fd::{AsFd, BorrowedFd, RawFd}, str::Chars};

use nix::{errno::Errno, libc::{self, STDIN_FILENO}, poll::{self, PollFlags, PollTimeout}, sys::termios, unistd::isatty};
use unicode_segmentation::UnicodeSegmentation;
use unicode_width::{UnicodeWidthChar, UnicodeWidthStr};

use crate::{libsh::error::{ShErr, ShErrKind, ShResult}, prompt::readline::keys::{KeyCode, ModKeys}};
use crate::prelude::*;

use super::{keys::KeyEvent, linebuf::LineBuf};

pub fn raw_mode() -> RawModeGuard {
	let orig = termios::tcgetattr(unsafe{BorrowedFd::borrow_raw(STDIN_FILENO)}).expect("Failed to get terminal attributes");
	let mut raw = orig.clone();
	termios::cfmakeraw(&mut raw);
	termios::tcsetattr(unsafe{BorrowedFd::borrow_raw(STDIN_FILENO)}, termios::SetArg::TCSANOW, &raw)
		.expect("Failed to set terminal to raw mode");
	RawModeGuard { orig, fd: STDIN_FILENO }
}

pub type Row = u16;
pub type Col = u16;

#[derive(Default,Clone,Copy,PartialEq,Eq,PartialOrd,Ord,Debug)]
pub struct Pos {
	col: Col,
	row: Row
}

// I'd like to thank rustyline for this idea
nix::ioctl_read_bad!(win_size, libc::TIOCGWINSZ, libc::winsize);

pub fn get_win_size(fd: RawFd) -> (Col,Row) {
	use std::mem::zeroed;

	if cfg!(test) {
		return (80,24)
	}

	unsafe {
		let mut size: libc::winsize = zeroed();
		match win_size(fd, &mut size) {
			Ok(0) => {
				/* rustyline code says:
				 In linux pseudo-terminals are created with dimensions of
				 zero. If host application didn't initialize the correct
				 size before start we treat zero size as 80 columns and
				 infinite rows
				*/
				let cols = if size.ws_col == 0 { 80 } else { size.ws_col };
				let rows = if size.ws_row == 0 {
					u16::MAX
				} else {
					size.ws_row
				};
				(cols, rows)
			}
			_ => (80,24)
		}
	}
}

fn write_all(fd: RawFd, buf: &str) -> nix::Result<()> {
	let mut bytes = buf.as_bytes();
	while !bytes.is_empty() {
		match nix::unistd::write(unsafe { BorrowedFd::borrow_raw(fd) }, bytes) {
			Ok(0) => return Err(Errno::EIO),
			Ok(n) => bytes = &bytes[n..],
			Err(Errno::EINTR) => {}
			Err(r) => return Err(r),
		}
	}
	Ok(())
}

// Big credit to rustyline for this
fn width(s: &str, esc_seq: &mut u8) -> u16 {
	let w_calc = width_calculator();
	if *esc_seq == 1 {
		if s == "[" {
			// CSI
			*esc_seq = 2;
		} else {
			// two-character sequence
			*esc_seq = 0;
		}
		0
	} else if *esc_seq == 2 {
		if s == ";" || (s.as_bytes()[0] >= b'0' && s.as_bytes()[0] <= b'9') {
			/*} else if s == "m" {
			// last
			 *esc_seq = 0;*/
	} else {
		// not supported
		*esc_seq = 0;
	}
	0
	} else if s == "\x1b" {
		*esc_seq = 1;
		0
	} else if s == "\n" {
		0
	} else {
		w_calc.width(s) as u16
	}
}

pub fn width_calculator() -> Box<dyn WidthCalculator> {
	match env::var("TERM_PROGRAM").as_deref() {
		Ok("Apple_Terminal") => Box::new(UnicodeWidth),
		Ok("iTerm.app") => Box::new(UnicodeWidth),
		Ok("WezTerm") => Box::new(UnicodeWidth),
		Err(std::env::VarError::NotPresent) => match std::env::var("TERM").as_deref() {
			Ok("xterm-kitty") => Box::new(NoZwj),
			_ => Box::new(WcWidth)
		},
		_ => Box::new(WcWidth)
	}
}

fn read_digits_until(rdr: &mut TermReader, sep: char) -> ShResult<Option<u32>> {
	let mut num: u32 = 0;
	loop {
		match rdr.next_byte()? as char {
			digit @ '0'..='9' => {
				let digit = digit.to_digit(10).unwrap();
				num = append_digit(num, digit);
				continue;
			}
			c if c == sep => break,
			_ => return Ok(None),
		}
	}
	Ok(Some(num))
}

pub fn append_digit(left: u32, right: u32) -> u32 {
	left.saturating_mul(10)
		.saturating_add(right)
}


pub trait WidthCalculator {
	fn width(&self, text: &str) -> usize;
}

pub trait KeyReader {
	fn read_key(&mut self) -> Option<KeyEvent>;
}

pub trait LineWriter {
	fn clear_rows(&mut self, layout: &Layout) -> ShResult<()>;
	fn redraw(
		&mut self,
		prompt: &str,
		line: &LineBuf,
		new_layout: &Layout,
	) -> ShResult<()>;
	fn flush_write(&mut self, buf: &str) -> ShResult<()>;
}

#[derive(Clone,Copy,Debug)]
pub struct UnicodeWidth;

impl WidthCalculator for UnicodeWidth {
	fn width(&self, text: &str) -> usize {
		text.width()
	}
}

#[derive(Clone,Copy,Debug)]
pub struct WcWidth;

impl WcWidth {
	pub fn cwidth(&self, ch: char) -> usize {
		ch.width().unwrap()
	}
}

impl WidthCalculator for WcWidth {
	fn width(&self, text: &str) -> usize {
		let mut width = 0;
		for ch in text.chars() {
			width += self.cwidth(ch)
		}
		width
	}
}

const ZWJ: char = '\u{200D}';
#[derive(Clone,Copy,Debug)]
pub struct NoZwj;

impl WidthCalculator for NoZwj {
	fn width(&self, text: &str) -> usize {
		let mut width = 0;
		for slice in text.split(ZWJ) {
			width += UnicodeWidth.width(slice);
		}
		width
	}
}

pub struct TermBuffer {
	tty: RawFd
}

impl TermBuffer {
	pub fn new(tty: RawFd) -> Self {
		assert!(isatty(tty).is_ok_and(|r| r));
		Self {
			tty
		}
	}
}

impl Read for TermBuffer {
	fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
		assert!(isatty(self.tty).is_ok_and(|r| r));
		loop {
			match nix::unistd::read(self.tty, buf) {
				Ok(n) => return Ok(n),
				Err(Errno::EINTR) => {}
				Err(e) => return Err(std::io::Error::from_raw_os_error(e as i32))
			}
		}
	}
}

pub struct RawModeGuard {
	orig: termios::Termios,
	fd: RawFd,
}

impl RawModeGuard {
	/// Disable raw mode temporarily for a specific operation
	pub fn disable_for<F: FnOnce() -> R, R>(&self, func: F) -> R {
		unsafe {
			let fd = BorrowedFd::borrow_raw(self.fd);
			// Temporarily restore the original termios
			termios::tcsetattr(fd, termios::SetArg::TCSANOW, &self.orig)
				.expect("Failed to temporarily disable raw mode");

			// Run the function
			let result = func();

			// Re-enable raw mode
			let mut raw = self.orig.clone();
			termios::cfmakeraw(&mut raw);
			termios::tcsetattr(fd, termios::SetArg::TCSANOW, &raw)
				.expect("Failed to re-enable raw mode");

			result
		}
	}
}

impl Drop for RawModeGuard {
	fn drop(&mut self) {
		unsafe { 
			let _ = termios::tcsetattr(BorrowedFd::borrow_raw(self.fd), termios::SetArg::TCSANOW, &self.orig);
		}
	}
}

pub struct TermReader {
	buffer: BufReader<TermBuffer>
}

impl Default for TermReader {
	fn default() -> Self {
	  Self::new()
	}
}

impl TermReader {
	pub fn new() -> Self {
		Self {
			buffer: BufReader::new(TermBuffer::new(1))
		}
	}


	/// Execute some logic in raw mode
	/// 
	/// Saves the termios before running the given function.
	/// If the given function panics, the panic will halt momentarily to restore the termios
	pub fn poll(&mut self, timeout: PollTimeout) -> ShResult<bool> {
		if !self.buffer.buffer().is_empty() {
			return Ok(true)
		}

		let mut fds = [poll::PollFd::new(self.as_fd(),PollFlags::POLLIN)];
		let r = poll::poll(&mut fds, timeout);
		match r {
			Ok(n) => Ok(n != 0),
			Err(Errno::EINTR) => Ok(false),
			Err(e) => Err(e.into())
		}
	}

	pub fn next_byte(&mut self) -> std::io::Result<u8> {
		let mut buf = [0u8];
		self.buffer.read_exact(&mut buf)?;
		Ok(buf[0])
	}

	pub fn peek_byte(&mut self) -> std::io::Result<u8> {
		let buf = self.buffer.fill_buf()?;
		if buf.is_empty() {
			Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "EOF"))
		} else {
			Ok(buf[0])
		}
	}

	pub fn consume_byte(&mut self) {
		self.buffer.consume(1);
	}



	pub fn parse_esc_seq(&mut self) -> ShResult<KeyEvent> {
		let mut seq = vec![0x1b];

		let b1 = self.peek_byte()?;
		self.consume_byte();
		seq.push(b1);

		match b1 {
			b'[' => {
				let b2 = self.peek_byte()?;
				self.consume_byte();
				seq.push(b2);

				match b2 {
					b'A' => Ok(KeyEvent(KeyCode::Up, ModKeys::empty())),
					b'B' => Ok(KeyEvent(KeyCode::Down, ModKeys::empty())),
					b'C' => Ok(KeyEvent(KeyCode::Right, ModKeys::empty())),
					b'D' => Ok(KeyEvent(KeyCode::Left, ModKeys::empty())),
					b'1'..=b'9' => {
						let mut digits = vec![b2];

						loop {
							let b = self.peek_byte()?;
							seq.push(b);
							self.consume_byte();

							if b == b'~' || b == b';' {
								break;
							} else if b.is_ascii_digit() {
								digits.push(b);
							} else {
								break;
							}
						}

						let key = match digits.as_slice() {
							[b'1'] => KeyCode::Home,
							[b'3'] => KeyCode::Delete,
							[b'4'] => KeyCode::End,
							[b'5'] => KeyCode::PageUp,
							[b'6'] => KeyCode::PageDown,
							[b'7'] => KeyCode::Home, // xterm alternate
							[b'8'] => KeyCode::End,  // xterm alternate

							[b'1', b'5'] => KeyCode::F(5),
							[b'1', b'7'] => KeyCode::F(6),
							[b'1', b'8'] => KeyCode::F(7),
							[b'1', b'9'] => KeyCode::F(8),
							[b'2', b'0'] => KeyCode::F(9),
							[b'2', b'1'] => KeyCode::F(10),
							[b'2', b'3'] => KeyCode::F(11),
							[b'2', b'4'] => KeyCode::F(12),
							_ => KeyCode::Esc,
						};

						Ok(KeyEvent(key, ModKeys::empty()))
					}
					_ => Ok(KeyEvent(KeyCode::Esc, ModKeys::empty())),
				}
			}
			b'O' => {
				let b2 = self.peek_byte()?;
				self.consume_byte();
				seq.push(b2);

				let key = match b2 {
					b'P' => KeyCode::F(1),
					b'Q' => KeyCode::F(2),
					b'R' => KeyCode::F(3),
					b'S' => KeyCode::F(4),
					_ => KeyCode::Esc,
				};

				Ok(KeyEvent(key, ModKeys::empty()))
			}
			_ => Ok(KeyEvent(KeyCode::Esc, ModKeys::empty())),
		}
	}

}

impl KeyReader for TermReader {
	fn read_key(&mut self) -> Option<KeyEvent> {
		use core::str;

		let mut collected = Vec::with_capacity(4);

		loop {
			let byte = self.next_byte().ok()?;
			flog!(DEBUG, "read byte: {:?}",byte as char);
			collected.push(byte);

			// If it's an escape seq, delegate to ESC sequence handler
			if collected[0] == 0x1b && collected.len() == 1 && self.poll(PollTimeout::ZERO).ok()? {
				return self.parse_esc_seq().ok();
			}

			// Try parse as valid UTF-8
			if let Ok(s) = str::from_utf8(&collected) {
				return Some(KeyEvent::new(s, ModKeys::empty()));
			}

			// UTF-8 max 4 bytes — if it’s invalid at this point, bail
			if collected.len() >= 4 {
				break;
			}
		}

		None
	}
}

impl AsFd for TermReader {
	fn as_fd(&self) -> BorrowedFd<'_> {
		let fd = self.buffer.get_ref().tty;
		unsafe { BorrowedFd::borrow_raw(fd) }
	}
}

pub struct Layout {
	pub w_calc: Box<dyn WidthCalculator>,
	pub prompt_end: Pos,
	pub cursor: Pos,
	pub end: Pos
}

impl Debug for Layout {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		writeln!(f, "Layout: ")?;
		writeln!(f, "\tPrompt End: {:?}",self.prompt_end)?;
		writeln!(f, "\tCursor: {:?}",self.cursor)?;
		writeln!(f, "\tEnd: {:?}",self.end)
	}
}

impl Layout {
	pub fn new() -> Self {
		let w_calc = width_calculator();
		Self {
			w_calc,
			prompt_end: Pos::default(),
			cursor: Pos::default(),
			end: Pos::default(),
		}
	}
	pub fn from_parts(
		tab_stop: u16,
		term_width: u16,
		prompt: &str,
		to_cursor: &str,
		to_end: &str,
	) -> Self {
		flog!(DEBUG,to_cursor);
		let prompt_end = Self::calc_pos(tab_stop, term_width,	prompt, Pos { col: 0, row: 0 });
		let cursor = Self::calc_pos(tab_stop, term_width, to_cursor, prompt_end);
		let end = Self::calc_pos(tab_stop, term_width, to_end, prompt_end);
		Layout { w_calc: width_calculator(), prompt_end, cursor, end }
	}

	pub fn calc_pos(tab_stop: u16, term_width: u16, s: &str, orig: Pos) -> Pos {
		let mut pos = orig;
		let mut esc_seq = 0;
		for c in s.graphemes(true) {
			if c == "\n" {
				pos.row += 1;
				pos.col = 0;
			}
			let c_width = if c == "\t" {
				tab_stop - (pos.col % tab_stop)
			} else {
				width(c, &mut esc_seq)
			};
			pos.col += c_width;
			if pos.col > term_width {
				pos.row += 1;
				pos.col = c_width;
			}
		}
		if pos.col >= term_width {
			pos.row += 1;
			pos.col = 0;
		}

		pos
	}
}

impl Default for Layout {
	fn default() -> Self {
	  Self::new()
	}
}

pub struct TermWriter {
	out: RawFd,
	t_cols: Col, // terminal width
	buffer: String,
	w_calc: Box<dyn WidthCalculator>,
	tab_stop: u16,
}

impl TermWriter {
	pub fn new(out: RawFd) -> Self {
		let w_calc = width_calculator();
		let (t_cols,_) = get_win_size(out);
		Self {
			out,
			t_cols,
			buffer: String::new(),
			w_calc,
			tab_stop: 8 // TODO: add a way to configure this
		}
	}
	pub fn get_cursor_movement(&self, old: Pos, new: Pos) -> ShResult<String> {
		let mut buffer = String::new();
		let err = |_| ShErr::simple(ShErrKind::InternalErr, "Failed to write to cursor movement buffer");

		match new.row.cmp(&old.row) {
			std::cmp::Ordering::Greater => {
				let shift = new.row - old.row;
				match shift {
					1 => buffer.push_str("\x1b[B"),
					_ => write!(buffer, "\x1b[{shift}B").map_err(err)?
				}
			}
			std::cmp::Ordering::Less => {
				let shift = old.row - new.row;
				match shift {
					1 => buffer.push_str("\x1b[A"),
					_ => write!(buffer, "\x1b[{shift}A").map_err(err)?
				}
			}
			std::cmp::Ordering::Equal => { /* Do nothing */ }
		}

		match new.col.cmp(&old.col) {
			std::cmp::Ordering::Greater => {
				let shift = new.col - old.col;
				match shift {
					1 => buffer.push_str("\x1b[C"),
					_ => write!(buffer, "\x1b[{shift}C").map_err(err)?
				}
			}
			std::cmp::Ordering::Less => {
				let shift = old.col - new.col;
				match shift {
					1 => buffer.push_str("\x1b[D"),
					_ => write!(buffer, "\x1b[{shift}D").map_err(err)?
				}
			}
			std::cmp::Ordering::Equal => { /* Do nothing */ }
		}
		Ok(buffer)
	}
	pub fn move_cursor(&mut self, old: Pos, new: Pos) -> ShResult<()> {
		self.buffer.clear();
		let movement = self.get_cursor_movement(old, new)?;

		write_all(self.out, &movement)?;
		Ok(())
	}



	pub fn update_t_cols(&mut self) {
		let (t_cols,_) = get_win_size(self.out);
		self.t_cols = t_cols;
	}

	pub fn move_cursor_at_leftmost(&mut self, rdr: &mut TermReader, use_newline: bool) -> ShResult<()> {
		let result = rdr.poll(PollTimeout::ZERO)?;
		if result {
			// The terminals reply is going to be stuck behind the currently buffered output
			// So let's get out of here
			return Ok(()) 
		}

		// Ping the cursor's position
		self.flush_write("\x1b[6n\n")?;

		if !rdr.poll(PollTimeout::from(255u8))? {
				return Ok(())
		}

		if rdr.next_byte()? as char != '\x1b' {
				return Ok(())
		}

		if rdr.next_byte()? as char != '[' {
				return Ok(())
		}

		if read_digits_until(rdr, ';')?.is_none() {
				return Ok(())
		}

		// We just consumed everything up to the column number, so let's get that now
		let col = read_digits_until(rdr, 'R')?;

		// The cursor is not at the leftmost, so let's fix that
		if col != Some(1) {
			if use_newline {
				// We use '\n' instead of '\r' sometimes because if there's a bunch of garbage on this line,
				// It might pollute the prompt/line buffer if those are shorter than said garbage
				self.flush_write("\n")?;
			} else {
				// Sometimes though, we know that there's nothing to the right of the cursor after moving
				// So we just move to the left.
				self.flush_write("\r")?;
			}
		}

		Ok(())
	}
}

impl LineWriter for TermWriter {
	fn clear_rows(&mut self, layout: &Layout) -> ShResult<()> {
		self.buffer.clear();
		let rows_to_clear = layout.end.row;
		let cursor_row = layout.cursor.row;

		let cursor_motion = rows_to_clear.saturating_sub(cursor_row);
		if cursor_motion > 0 {
			write!(self.buffer, "\x1b[{cursor_motion}B").unwrap()
		}

		for _ in 0..rows_to_clear {
			self.buffer.push_str("\x1b[2K\x1b[A");
		}
		self.buffer.push_str("\x1b[2K");
		write_all(self.out,self.buffer.as_str())?;
		self.buffer.clear();
		Ok(())
	}

	fn redraw(
		&mut self,
		prompt: &str,
		line: &LineBuf,
		new_layout: &Layout,
	) -> ShResult<()> {
		let err = |_| ShErr::simple(ShErrKind::InternalErr, "Failed to write to LineWriter internal buffer");
		self.buffer.clear();

		let end = new_layout.end;
		let cursor = new_layout.cursor;

		self.buffer.push_str(prompt);
		self.buffer.push_str(line.as_str());

		if end.col == 0 && end.row > 0 && !self.buffer.ends_with('\n') {
			// The line has wrapped. We need to use our own line break.
			self.buffer.push('\n');
		}

		let movement = self.get_cursor_movement(end, cursor)?;
		write!(self.buffer, "{}", &movement).map_err(err)?;

		write_all(self.out, self.buffer.as_str())?;
		Ok(())
	}

	fn flush_write(&mut self, buf: &str) -> ShResult<()> {
		write_all(self.out, buf)?;
		Ok(())
	}
}
