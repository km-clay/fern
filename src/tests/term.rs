use libsh::term::{Style, StyleSet, Styled};

use super::super::*;
#[test]
fn styled_simple() {
	let input = "hello world";
	let styled = input.styled(Style::Green);

	insta::assert_snapshot!(styled)
}
#[test]
fn styled_multiple() {
	let input = "styled text";
	let styled = input.styled(Style::Red | Style::Bold | Style::Underline);
	insta::assert_snapshot!(styled);
}
#[test]
fn styled_rgb() {
	let input = "RGB styled text";
	let styled = input.styled(Style::RGB(255, 99, 71));  // Tomato color
	insta::assert_snapshot!(styled);
}
#[test]
fn styled_background() {
	let input = "text with background";
	let styled = input.styled(Style::BgBlue | Style::Bold);
	insta::assert_snapshot!(styled);
}
#[test]
fn styled_set() {
	let input = "multi-style text";
	let style_set = StyleSet::new().add_style(Style::Magenta).add_style(Style::Italic);
	let styled = input.styled(style_set);
	insta::assert_snapshot!(styled);
}
#[test]
fn styled_reset() {
	let input = "reset test";
	let styled = input.styled(Style::Bold | Style::Reset);
	insta::assert_snapshot!(styled);
}
