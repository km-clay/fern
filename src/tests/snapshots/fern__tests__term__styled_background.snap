---
source: src/tests/term.rs
expression: styled
---
[44m[1mtext with background[0m
